library/ecp.c

```diff
@@ -2429,6 +2434,306 @@ static int ecp_mul_mxz( mbedtls_ecp_group *grp, mbedtls_ecp_point *R,
+    mbedtls_mpi_init( &D ); mbedtls_mpi_init( &E ); mbedtls_mpi_init( &F );
+    mbedtls_mpi_init( &J );
+
+    /* B = (X1+Y1)^2 */
+    MBEDTLS_MPI_CHK( mbedtls_mpi_add_mod( grp, &B,    &P->X,   &P->Y ) );
+    MBEDTLS_MPI_CHK( mbedtls_mpi_mul_mod( grp, &B,    &B,      &B    ) );
+    /* C = X1^2 */
+    MBEDTLS_MPI_CHK( mbedtls_mpi_mul_mod( grp, &C,    &P->X,   &P->X ) );
+    /* D = Y1^2 */
+    MBEDTLS_MPI_CHK( mbedtls_mpi_mul_mod( grp, &D,    &P->Y,   &P->Y ) );
+    /* E = a*C */
+    MBEDTLS_MPI_CHK( mbedtls_mpi_mul_mod( grp, &E,    &grp->A, &C    ) );
+    /* F = E+D */
+    MBEDTLS_MPI_CHK( mbedtls_mpi_add_mod( grp, &F,    &E,      &D    ) );
+    /* J = F-2*(Z1^2) */
+    MBEDTLS_MPI_CHK( mbedtls_mpi_mul_mod( grp, &J,    &P->Z,   &P->Z ) );
+    MBEDTLS_MPI_CHK( mbedtls_mpi_shift_l_mod( grp, &J, 1             ) );
+    MBEDTLS_MPI_CHK( mbedtls_mpi_sub_mod( grp, &J,    &F,      &J    ) );
+    /* X3 = (B-C-D)*J */
+    MBEDTLS_MPI_CHK( mbedtls_mpi_sub_mod( grp, &R->X, &B,      &C    ) );
+    MBEDTLS_MPI_CHK( mbedtls_mpi_sub_mod( grp, &R->X, &R->X,   &D    ) );
+    MBEDTLS_MPI_CHK( mbedtls_mpi_mul_mod( grp, &R->X, &R->X,   &J    ) );
+    /* Y3 = F*(E-D) */
+    MBEDTLS_MPI_CHK( mbedtls_mpi_sub_mod( grp, &R->Y, &E,      &D    ) );
+    MBEDTLS_MPI_CHK( mbedtls_mpi_mul_mod( grp, &R->Y, &R->Y,   &F    ) );
+    /* Z3 = F*J */
+    MBEDTLS_MPI_CHK( mbedtls_mpi_mul_mod( grp, &R->Z, &F,      &J    ) );
+
+cleanup:
+    mbedtls_mpi_free( &A ); mbedtls_mpi_free( &B ); mbedtls_mpi_free( &C );
+    mbedtls_mpi_free( &D ); mbedtls_mpi_free( &E ); mbedtls_mpi_free( &F );
+    mbedtls_mpi_free( &J );
+
+    return( ret );
+}
+
+/*
+ * Multiplication with Montgomery ladder in x/y/z coordinates,
+ * for curves in Edwards form.
+ */
+static int ecp_mul_edxyz( mbedtls_ecp_group *grp, mbedtls_ecp_point *R,
+                          const mbedtls_mpi *m, const mbedtls_ecp_point *P,
+                          int (*f_rng)(void *, unsigned char *, size_t),
+                          void *p_rng )
+{
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    size_t i;
+    unsigned char b;
+    mbedtls_ecp_point RP;
+
+    mbedtls_ecp_point_init( &RP );
+
+    /* Read from P before writing to R, in case P == R */
+    MBEDTLS_MPI_CHK( mbedtls_ecp_copy( &RP, P ) );
+
+    /* Set R to zero in projective coordinates */
+    MBEDTLS_MPI_CHK( mbedtls_mpi_lset( &R->X, 0 ) );
+    MBEDTLS_MPI_CHK( mbedtls_mpi_lset( &R->Y, 1 ) );
+    MBEDTLS_MPI_CHK( mbedtls_mpi_lset( &R->Z, 1 ) );
+
+    /* RP.X and RP.Y might be slightly larger than P, so reduce them */
+    MOD_ADD( RP.X );
+    MOD_ADD( RP.Y );
+
+    /* Randomize coordinates of the starting point */
+    if( f_rng != NULL )
+        MBEDTLS_MPI_CHK( ecp_randomize_edxyz( grp, &RP, f_rng, p_rng ) );
+
+    /* Loop invariant: R = result so far, RP = R + P */
+    i = mbedtls_mpi_bitlen( m ); /* one past the (zero-based) most significant bit */
```

Comment:
This leaks the bit length of m through timing. I'm not sure about EdDSA off-hand, but for ECDSA this is fatal, see for example the "[Remote Timing Attacks are Still Practical](Remote Timing Attacks are Still Practical)" paper. I'd be much more comfortable setting `i` to a known value that depends only on `grp`. Or alternatively, add a comment with a very strong argument why leaking the bit-length of `m` is never an issue - but not leaking it really would be better.